src/os/unix/wayland.rs

```diff
@@ -0,0 +1,1182 @@
+            *touch_fl.borrow(),
+        );
+
+        ret
+    }
+}
+
+struct WaylandInput {
+    kb_events: mpsc::Receiver<wayland_client::protocol::wl_keyboard::Event>,
+    pt_events: mpsc::Receiver<wayland_client::protocol::wl_pointer::Event>,
+    input_devs: (Main<WlKeyboard>, Main<WlPointer>),
+}
+
+impl WaylandInput {
+    fn new(seat: &Main<WlSeat>) -> Self {
+        let (keyboard, pointer, _touch) = (seat.get_keyboard(), seat.get_pointer(), ());
+
+        let (kb_sender, kb_receiver) = mpsc::sync_channel(1024);
+
+        keyboard.quick_assign(move |_, event, _| {
+            kb_sender.send(event).unwrap();
+        });
+
+        let (pt_sender, pt_receiver) = mpsc::sync_channel(1024);
+
+        pointer.quick_assign(move |_, event, _| {
+            pt_sender.send(event).unwrap();
+        });
+
+        Self {
+            kb_events: kb_receiver,
+            pt_events: pt_receiver,
+            input_devs: (keyboard, pointer),
+        }
+    }
+
+    fn get_pointer(&self) -> &Main<WlPointer> {
+        &self.input_devs.1
+    }
+
+    fn iter_keyboard_events(
+        &self,
+    ) -> std::sync::mpsc::TryIter<wayland_client::protocol::wl_keyboard::Event> {
+        self.kb_events.try_iter()
+    }
+
+    fn iter_pointer_events(
+        &self,
+    ) -> std::sync::mpsc::TryIter<wayland_client::protocol::wl_pointer::Event> {
+        self.pt_events.try_iter()
+    }
+}
+
+pub struct Window {
+    display: DisplayInfo,
+
+    width: i32,
+    height: i32,
+
+    scale: i32,
+    bg_color: u32,
+    scale_mode: ScaleMode,
+
+    mouse_x: f32,
+    mouse_y: f32,
+    scroll_x: f32,
+    scroll_y: f32,
+    buttons: [u8; 3],
+    prev_cursor: CursorStyle,
+
+    should_close: bool,
+    active: bool,
+
+    key_handler: KeyHandler,
+    //option because MaybeUninit's get_ref() is nightly-only
+    keymap: Option<xkb::keymap::Keymap>,
+    update_rate: UpdateRate,
+    menu_counter: MenuHandle,
+    menus: Vec<UnixMenu>,
+    input: WaylandInput,
+    resizable: bool,
+    //temporary buffer
+    buffer: Vec<u32>,
+    //configure, close
+    toplevel_info: (Rc<RefCell<Option<(i32, i32)>>>, Rc<RefCell<bool>>),
+}
+
+impl Window {
+    pub fn new(name: &str, width: usize, height: usize, opts: WindowOptions) -> Result<Self> {
+        let scale: i32 = match opts.scale {
+            //STUB: currently not working in Wayland
+            Scale::FitScreen => 1,
+
+            Scale::X1 => 1,
+            Scale::X2 => 2,
+            Scale::X4 => 4,
+            Scale::X8 => 8,
+            Scale::X16 => 16,
+            Scale::X32 => 32,
+        };
+
+        let (dsp, input) = DisplayInfo::new(
+            (width as i32 * scale, height as i32 * scale),
+            false,
+            !opts.borderless,
+        )?;
+        if opts.title {
+            dsp.set_title(name);
+        }
+        if !opts.resize {
+            dsp.set_no_resize((width as i32 * scale, height as i32 * scale));
+        }
+
+        //TODO: maybe outsource
```

Comment:
Expand upon this comment, please.

Reply:
I thought about putting the initialization of "close" and "configure" into an own function.